### ISSUE TITLE
Memory leak in MPI_Allreduce when using a repeatedly created and freed MPI_Datatype

### ISSUE THREAD

**Issue by tamiko** [open-mpi/ompi#8827]
I observe a memory leak in `MPI_Allreduce` when using a repeatedly created and freed custom `MPI_Datatype`. This issue seems to first occur in ompi version 4.1.0. I cannot reproduce with ompi 4.0.5.

Minimimal source code (C++) demonstrating the problem attached.

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

4.1.0

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

ompi was installed via debian bullseye distribution manager.

```
# ompi_info
                 Package: Debian OpenMPI
                Open MPI: 4.1.0
  Open MPI repo revision: v4.1.0
   Open MPI release date: Dec 18, 2020
                Open RTE: 4.1.0
  Open RTE repo revision: v4.1.0
   Open RTE release date: Dec 18, 2020
                    OPAL: 4.1.0
      OPAL repo revision: v4.1.0
       OPAL release date: Dec 18, 2020
                 MPI API: 3.1.0
            Ident string: 4.1.0
                  Prefix: /usr
 Configured architecture: x86_64-pc-linux-gnu
          Configure host: debian
           Configured by: packaging
           Configured on: Wed Jan 20 17:22:18 UTC 2021
          Configure host: debian
  Configure command line: '--build=x86_64-linux-gnu' '--prefix=/usr' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-option-checking' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--runstatedir=/run' '--disable-maintainer-mode' '--disable-dependency-tracking' '--disable-silent-rules' '--disable-wrapper-runpath' '--with-package-string=Debian OpenMPI' '--with-verbs' '--with-libfabric' '--with-psm' '--with-psm2' '--with-ucx' '--with-pmix=/usr/lib/x86_64-linux-gnu/pmix2' '--with-jdk-dir=/usr/lib/jvm/default-java' '--enable-mpi-java' '--enable-opal-btl-usnic-unit-tests' '--with-libevent=external' '--with-hwloc=external' '--disable-silent-rules' '--enable-mpi-cxx' '--with-devel-headers' '--with-slurm' '--with-sge' '--without-tm' '--sysconfdir=/etc/openmpi' '--libdir=${prefix}/lib/x86_64-linux-gnu/openmpi/lib' '--includedir=${prefix}/lib/x86_64-linux-gnu/openmpi/include'
                Built by: 
                Built on: Wed Jan 20 17:22:18 UTC 2021
              Built host: debian
              C bindings: yes
            C++ bindings: yes
             Fort mpif.h: yes (all)
            Fort use mpi: yes (full: ignore TKR)
       Fort use mpi size: deprecated-ompi-info-value
        Fort use mpi_f08: yes
 Fort mpi_f08 compliance: The mpi_f08 module is available, but due to limitations in the gfortran compiler and/or Open MPI, does not support the following: array subsections, direct passthru (where possible) to underlying Open MPI's C functionality
  Fort mpi_f08 subarrays: no
           Java bindings: yes
  Wrapper compiler rpath: rpath
              C compiler: gcc
     C compiler absolute: /usr/bin/gcc
  C compiler family name: GNU
      C compiler version: 10.2.1
            C++ compiler: g++
   C++ compiler absolute: /usr/bin/g++
           Fort compiler: gfortran
       Fort compiler abs: /usr/bin/gfortran
         Fort ignore TKR: yes (!GCC$ ATTRIBUTES NO_ARG_CHECK ::)
   Fort 08 assumed shape: yes
      Fort optional args: yes
          Fort INTERFACE: yes
    Fort ISO_FORTRAN_ENV: yes
       Fort STORAGE_SIZE: yes
      Fort BIND(C) (all): yes
      Fort ISO_C_BINDING: yes
 Fort SUBROUTINE BIND(C): yes
       Fort TYPE,BIND(C): yes
 Fort T,BIND(C,name="a"): yes
            Fort PRIVATE: yes
          Fort PROTECTED: yes
           Fort ABSTRACT: yes
       Fort ASYNCHRONOUS: yes
          Fort PROCEDURE: yes
         Fort USE...ONLY: yes
           Fort C_FUNLOC: yes
 Fort f08 using wrappers: yes
         Fort MPI_SIZEOF: yes
             C profiling: yes
           C++ profiling: yes
   Fort mpif.h profiling: yes
  Fort use mpi profiling: yes
   Fort use mpi_f08 prof: yes
          C++ exceptions: no
          Thread support: posix (MPI_THREAD_MULTIPLE: yes, OPAL support: yes, OMPI progress: no, ORTE progress: yes, Event lib: yes)
           Sparse Groups: no
  Internal debug support: no
  MPI interface warnings: yes
     MPI parameter check: runtime
Memory profiling support: no
Memory debugging support: no
              dl support: yes
   Heterogeneous support: no
 mpirun default --prefix: no
       MPI_WTIME support: native
     Symbol vis. support: yes
   Host topology support: yes
            IPv6 support: no
      MPI1 compatibility: no
          MPI extensions: affinity, cuda, pcollreq
   FT Checkpoint support: no (checkpoint thread: no)
   C/R Enabled Debugging: no
  MPI_MAX_PROCESSOR_NAME: 256
    MPI_MAX_ERROR_STRING: 256
     MPI_MAX_OBJECT_NAME: 64
        MPI_MAX_INFO_KEY: 36
        MPI_MAX_INFO_VAL: 256
       MPI_MAX_PORT_NAME: 1024
  MPI_MAX_DATAREP_STRING: 128
```

### Please describe the system on which you are running

* Operating system/version: Debian testing (bullseye)
* Computer hardware: 2x Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz, 32 cores + HT, 512GB RAM
* Network type: vader BTL, (but in general infiniband via ofi)

-----------------------------

## Details of the problem

I observe a memory leak in `MPI_Allreduce` when using a repeatedly created and freed custom `MPI_Datatype`. This issue seems to first occur in ompi version 4.1.0. I cannot reproduce with ompi 4.0.5.

When running the followin example code with something like `mpirun -np 10` I observe a memory leak (continuously growing RSS).

I am aware that creating and releasing the same `MPI_Datatype` over and over again is not very elegant. But nevertheless the following code should not leak (and indeed it did not do so in previous ompi versions).

```cpp
#include <cstddef>
#include <limits>

#include <mpi.h>

void my_mpi_op(const void *, void *, MPI_Datatype *) {
  // do nothing
}

MPI_Op op;

void call_allreduce() {

  // create an MPI struct with 1 double:

  MPI_Datatype type;
  int          lengths[]       = {1};
  MPI_Aint     displacements[] = {0};
  MPI_Datatype types[]         = {MPI_DOUBLE};
  MPI_Type_create_struct(1, lengths, displacements, types, &type);

  MPI_Type_commit(&type);

  double in = 42.;
  double result = 0.;
  MPI_Allreduce(&in, &result, 1, type, op, MPI_COMM_WORLD);

  MPI_Type_free(&type);
}

int main(int argc, char *argv[]) {
  MPI_Init(&argc, &argv);

  MPI_Op_create(reinterpret_cast<MPI_User_function *>(&my_mpi_op), true, &op);

  while (true) {
    call_allreduce();
  }

  MPI_Op_free(&op);

  MPI_Finalize();
}
```

**Comment by yosefe** [comment]
@tamiko thank you for the report.
Can you please add "-mca pml ob1 -mca btl vader" as an experiment to see if UCX is causing the issue?
@hoopoepg i think we have leak in mca_pml_ucx_init_datatype: if the type is contig, we do not call "ompi_attr_set_c" to register type destructor. probably, the "out:" label from line 241 should be moved to line 228

**Comment by tamiko** [comment]
@yosefe I can confirm that `mpirun -mca pml ob1 -mca btl self,vader -np 10` does indeed not leak.

Also, I forgot to mention: Yes, I can confirm that the leak occurs in `mca_pml_ucx_init_datatype` (called over `MPI_Allreduce`). (I ran a heap memory profiler, "heaptrack", when investigating this issue.)

**Comment by yosefe** [comment]
@tamiko thank you for the confirmation,
Any chance you can try the following patch as a potential fix:
```diff
diff --git a/ompi/mca/pml/ucx/pml_ucx_datatype.c b/ompi/mca/pml/ucx/pml_ucx_datatype.c
index 5b1b8ccbed..98fa7145de 100644
--- a/ompi/mca/pml/ucx/pml_ucx_datatype.c
+++ b/ompi/mca/pml/ucx/pml_ucx_datatype.c
@@ -222,6 +222,7 @@ ucp_datatype_t mca_pml_ucx_init_datatype(ompi_datatype_t *datatype)
         ompi_mpi_abort(&ompi_mpi_comm_world.comm, 1);
     }
 
+out:
     /* Add custom attribute, to clean up UCX resources when OMPI datatype is
      * released.
      */
@@ -238,7 +239,6 @@ ucp_datatype_t mca_pml_ucx_init_datatype(ompi_datatype_t *datatype)
             ompi_mpi_abort(&ompi_mpi_comm_world.comm, 1);
         }
     }
-out:
     PML_UCX_VERBOSE(7, "created generic UCX datatype 0x%"PRIx64, ucp_datatype)
 
 #ifdef HAVE_UCP_REQUEST_PARAM_T
```

**Comment by tamiko** [comment]
@yosefe I was already on it :smile:  Yes, the above patch indeed fixes the issue!

**Comment by terwin** [comment]
I found this issue after identifying this same memory leak in my own application using 4.1.1. I believe this issue has been resolved by #8828, and it is safe to close this issue.
